### PR TITLE
FIX: Ensure first post is loaded before trying to bookmark topic

### PR DESCRIPTION
**Pull request by martin-brennan**
Load first post if not loaded when trying to bookmark topic

* if the first post was not loaded we could not bookmark it because
  now we call the toggleBookmarkReminder function on an actual post
  to open the modal window
* add a firstPost function to the topic model to get the first post
  from the stream if it is loaded and if not do a GET request to load
  it
* use the firstPost promise when bookmarking topic

Note: Way easier to review with whitespace disabled `https://github.com/discourse/discourse/pull/9382/files?w=1`

### CODE REVIEW

File: app/assets/javascripts/discourse/controllers/topic.js [hunk 1]
```diff
@@ -451,22 +451,11 @@ export default Controller.extend(bufferedProperty("model"), {
...
-            .loadPost(postId)
-            .then(post => {
-              this.send("editPost", post);
-            });
-        }
-      }
-      this.send("editPost", firstPost);
+      this.get("model")
+        .firstPost()
+        .then(firstPost => {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/controllers/topic.js:456]: can be one line

File: app/assets/javascripts/discourse/controllers/topic.js [hunk 2]
```diff
@@ -451,22 +451,11 @@ export default Controller.extend(bufferedProperty("model"), {
...
-        if (firstPost === undefined) {
-          return this.get("model.postStream")
-            .loadPost(postId)
-            .then(post => {
-              this.send("editPost", post);
-            });
-        }
-      }
-      this.send("editPost", firstPost);
+      this.get("model")
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/controllers/topic.js:454]: can't we just do `this.model` ?

File: app/assets/javascripts/discourse/models/topic.js [hunk 3]
```diff
@@ -407,82 +407,99 @@ const Topic = RestModel.extend({
     }
   },
 
+  firstPost() {
+    const postStream = this.postStream;
+    let firstPost = postStream.get("posts.firstObject");
+
+    if (firstPost.get("post_number") === 1) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/models/topic.js:414]:
```suggestion
    if (firstPost.post_number === 1) {
```

File: app/assets/javascripts/discourse/models/topic.js [hunk 4]
```diff
@@ -407,82 +407,99 @@ const Topic = RestModel.extend({
...
+              if (posts) {
+                const updated = [];
+                posts.forEach(post => {
+                  if (post.get("bookmarked")) {
+                    post.set("bookmarked", false);
+                    updated.push(post.id);
+                  }
+                  if (
+                    this.siteSettings.enable_bookmarks_with_reminders &&
+                    post.get("bookmarked_with_reminder")
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/models/topic.js:467]:
```suggestion
                    post.bookmarked_with_reminder
```

File: app/assets/javascripts/discourse/models/topic.js [hunk 5]
```diff
@@ -407,82 +407,99 @@ const Topic = RestModel.extend({
...
         } else {
-          return ajax(`/t/${this.id}/bookmark`, { type: "PUT" })
+          return ajax(`/t/${this.id}/remove_bookmarks`, { type: "PUT" })
             .then(() => {
               this.toggleProperty("bookmarked");
-              return this.afterTopicBookmarked(firstPost);
+              if (posts) {
+                const updated = [];
+                posts.forEach(post => {
+                  if (post.get("bookmarked")) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/models/topic.js:461]:
```suggestion
                  if (post.bookmarked) {
```

File: app/assets/javascripts/discourse/models/topic.js [hunk 6]
```diff
@@ -407,82 +407,99 @@ const Topic = RestModel.extend({
...
-          I18n.t("yes_value"),
-          confirmed =>
-            confirmed ? toggleBookmarkOnServer().then(resolve) : resolve()
+      };
+
+      const unbookmarkedPosts = [];
+      if (!bookmark && posts) {
+        posts.forEach(
+          post =>
+            (post.get("bookmarked") || post.get("bookmarked_with_reminder")) &&
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/models/topic.js:485]:
```suggestion
            (post.bookmarked || post.bookmarked_with_reminder) &&
```

File: app/assets/javascripts/discourse/models/topic.js [hunk 7]
```diff
@@ -407,82 +407,99 @@ const Topic = RestModel.extend({
...
+              .then(() => {
+                this.toggleProperty("bookmarked");
+                return this.afterTopicBookmarked(firstPost);
+              })
+              .catch(popupAjaxError)
+              .finally(() => this.set("bookmarking", false));
+          }
         } else {
-          return ajax(`/t/${this.id}/bookmark`, { type: "PUT" })
+          return ajax(`/t/${this.id}/remove_bookmarks`, { type: "PUT" })
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/models/topic.js:455]: I know this is not related to this PR, but shouldn't removing a bookmark be a DELETE request and not a PUT ?

File: app/assets/javascripts/discourse/models/topic.js [hunk 8]
```diff
@@ -407,82 +407,99 @@ const Topic = RestModel.extend({
...
     const bookmark = !this.bookmarked;
-
-    const toggleBookmarkOnServer = () => {
-      if (bookmark) {
-        if (this.siteSettings.enable_bookmarks_with_reminders) {
-          return firstPost.toggleBookmarkWithReminder().then(() => {
-            this.set("bookmarking", false);
-            return this.afterTopicBookmarked(firstPost);
-          });
+    let posts = get(this.postStream, "posts");
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/models/topic.js:435]: Any reason we don't do `this.postStream.posts` here ?

File: app/assets/javascripts/discourse/models/topic.js (hunk 7 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/models/topic.js:455]: Good point, will open another PR to fix this

**Pull request merged**